[Merged by Bors] - Create a local testnet

winksaville opened the pull request:
The testnet will be on the local computer and have 1 eth1 node,
4 beacon nodes, 1 validator with 20 vc's.

winksaville committed:
Create a local testnet

The testnet will be on the local computer and have 1 eth1 node,
4 beacon nodes, 1 validator with 20 vc's.

pawanjay176 commented on scripts/local_testnet/create_local_testnet.sh:
```diff
@@ -0,0 +1,60 @@
+execute_command_add_PID bootnode.log ./bootnode.sh
+
+sleeping 1
+execute_command_add_PID becacon_node_1.log ./beacon_node.sh $TESTNET_DIR/node_1 9001 8001
```
These need to be `$DATADIR/..` not `$TESTNET_DIR`

pawanjay176 commented on scripts/local_testnet/create_local_testnet.sh:
```diff
@@ -0,0 +1,60 @@
+sleeping 1
+execute_command_add_PID becacon_node_1.log ./beacon_node.sh $TESTNET_DIR/node_1 9001 8001
+sleeping 0
+execute_command_add_PID becacon_node_2.log ./beacon_node.sh $TESTNET_DIR/node_2 9002 8002
```
It would be great if we could pass the logging verbosity for the bn and vc to the `create_local_testnet.sh` script.
e.g `./create_local_testnet.sh debug` would get all debug logs.

Should be easy enough to do as logging verbosity is an optional parameter to both the `beacon_node.sh` and `validator_client.sh` scripts.

pawanjay176 reviewed (changes requested):
This is awesome! thanks for doing this :)

Just a few small changes.

winksaville committed:
Start a local testnet

The default is a the testnet will have 1 eth1 node, 4 beacon nodes,
with 20 validators per node and 1 validator client using the 20
associated with beacon node 1.

winksaville commented:
> This is awesome! thanks for doing this :)
> 
> Just a few small changes.

@pawanjay176, I made your suggested changes and tweaked a few other things plus added some initial documentation, txs for the review!

pawanjay176 commented on scripts/local_testnet/README.md:
````diff
@@ -15,45 +15,27 @@ make install-lcli
-```bash
-./setup.sh
-```
+Start a local eth1 ganache server plus boot node along with NODE_COUNT
````
```suggestion
Start a local eth1 ganache server plus boot node along with `NODE_COUNT`
```

winksaville replied:
Done, but changed to `BN_COUNT`

pawanjay176 commented on scripts/local_testnet/README.md:
````diff
@@ -15,45 +15,27 @@ make install-lcli
-```bash
-./bootnode.sh
-```
+This script takes two options `-c VC_COUNT` and `-d DEBUG_LEVEL`.
````
```suggestion
The `start_local_testnet.sh` script takes two options `-c VC_COUNT` and `-d DEBUG_LEVEL`.
```

winksaville replied:
Done, but changed `-c` to `-v`

pawanjay176 commented on scripts/local_testnet/README.md:
````diff
@@ -15,45 +15,27 @@ make install-lcli
-```
+This script takes two options `-c VC_COUNT` and `-d DEBUG_LEVEL`.
+The options maybe in any order or absent in which case they take the default value specified.
+- VC_COUNT: the number of validator clients to create, default: `1`
````
I think the default here should be `VC_COUNT == NODE_COUNT` as we are distributing the validators among all nodes equally. 
It might also be easier to have this param in `vars.env`.
Also, having just 1/4th of the validators active would prevent finality, so we wouldn't want that to be default.

winksaville replied:
Done, I've add VC_COUNT to vars.env and default it to BN_COUNT.

pawanjay176 commented on scripts/local_testnet/beacon_node.sh:
```diff
@@ -13,6 +13,7 @@ DEBUG_LEVEL=${4:-info}
 exec lighthouse \
 	--debug-level $DEBUG_LEVEL \
 	bn \
+	--subscribe-all-subnets \
```
Not sure why this option was added. I don't think this should be present by default.

winksaville replied:
@michaelsproul [suggested I do that](https://discord.com/channels/605577013327167508/710978433823277099/884952367102046288) as it was needed when only using one `vc`. I've made it option `-s`.

pawanjay176 commented on scripts/local_testnet/start_local_testnet.sh:
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env bash
+# Start all processes necessary to create a
+# local testnet with 1 eth1 client,
+# 4 beacon nodes and 1 validator with 20 vc's.
```
```suggestion
# `NODE_COUNT` beacon nodes and `VC_COUNT` validators with `VALIDATOR_COUNT / NODE_COUNT` vc's.
```

winksaville replied:
Removed this and added the info to the `-h` Option.

pawanjay176 commented on scripts/local_testnet/vars.env:
```diff
@@ -18,7 +18,10 @@ GENESIS_VALIDATOR_COUNT=80
+# Number of seconds to delay to start genesis block.
+# If started by a script this can be 0, if starting by hand
+# use something like 180.
+GENESIS_DELAY=0
```
I think this should be >0 by default (60secs maybe?) as we usually also want to check genesis as well when running local testnets.

winksaville replied:
`GENESIS_DELAY=0` appears to work and long values just forces the user to wait longer before "regular" work/output from BN's and VC's. I could make it an option to `start_local_testnet.sh` and make have it default to 0 or a small number if 0 is wrong. The simplest option is to leave it at the some minimum value, 0, and when a developer needs to "check genesis" they can change `vars.env`.

What would you like to do?

michaelsproul replied:
I'm happy with genesis delay 0, tweaking vars.env is always an option.

pawanjay176 commented on scripts/local_testnet/README.md:
````diff
@@ -15,45 +15,27 @@ make install-lcli
-./setup.sh
-```
+Start a local eth1 ganache server plus boot node along with NODE_COUNT
+number of beacon nodes and VC_COUNT validator clients.
````
```suggestion
number of beacon nodes and `VC_COUNT` validator clients.
```

winksaville replied:
Done

pawanjay176 reviewed (changes requested):
A few more nits.

Also, I think we should still keep the instructions to run the individual scripts (setup, bootnode, ganache, beacon, validator) after the instructions to run with `start_local_testnet`

winksaville committed:
Change requested by Pawanjay

README.md
 - Added back the original manual creation docs
 - Some cleanup

beacon_node.sh
 - Add command line Options so as to allow --subscribe-all-subnets to be
   controlled by caller
 - Handle positional parameters using $ORIND from getopts

setup.sh
 -Rename NODE_COUNT to BN_COUNT

start_local_testnet.sh
 - rename -c to -v for VC_COUNT
 - Clean up -h
 - Use VC_COUNT and BN_COUNT
 - Use bn as for loop variable

winksaville reviewed (commented):
@pawanjay176, I believe I've address all of your comments except setting GENESIS_DELAY to zero, see my comment.

On a similar note, in `start_local_testnet.sh` I've got 3 `sleeping x` function calls and this is very fragile, what would you suggest as an alternatives solution?

winksaville commented:
@pawanjay176 @michaelsproul do either of you or anyone else have any other suggestions for changes?

One outstanding issue is what to do about GENESIS_DELAY, see a small discussion starts [here](https://github.com/sigp/lighthouse/pull/2614#discussion_r715665354).

michaelsproul commented on scripts/local_testnet/README.md:
```diff
@@ -15,6 +15,32 @@ make install-lcli
+number of beacon nodes and `VC_COUNT` validator clients.
+
+The `start_local_testnet.sh` script takes three options `-v VC_COUNT`, `-d DEBUG_LEVEL` and `-h` for help.
+The options maybe in any order or absent in which case they take the default value specified.
```
```suggestion
The options may be in any order or absent in which case they take the default value specified.
```

winksaville replied:
Done

michaelsproul commented on scripts/local_testnet/README.md:
````diff
@@ -15,6 +15,32 @@ make install-lcli
+./start_local_testnet.sh
+```
+
+## Stoping the testnet
````
```suggestion
## Stopping the testnet
```

winksaville replied:
Done

michaelsproul commented on scripts/local_testnet/README.md:
```diff
@@ -15,6 +15,32 @@ make install-lcli
+
+## Manual creation of local testnet
+
+These scripts used by ./start_local_testnet.sh and maybe used to manually
```
```suggestion
These scripts are used by ./start_local_testnet.sh and may be used manually
```

winksaville replied:
Done

michaelsproul commented on scripts/local_testnet/beacon_node.sh:
```diff
@@ -1,25 +1,56 @@
 #
-# Usage: ./beacon_node.sh <DATADIR> <NETWORK-PORT> <HTTP-PORT> <OPTIONAL-DEBUG-LEVEL>
+
+# set -x
```
```suggestion

```

winksaville replied:
Done

michaelsproul reviewed (changes requested):
Looks good!

Tested it locally and I'm happy with it. Happy to merge once typos are corrected

winksaville committed:
Incorporate review changes from Michael Sproul

michaelsproul reviewed (approved):
Awesome, let's merge!

michaelsproul commented:
bors r+

michaelsproul commented:
The doppelganger protection tests are failing due to the BN_COUNT rename. We can merge once that's fixed, but I'm going to drop this PR from the current batch (and the v2.0.0-rc.0)

michaelsproul commented:
bors r-

winksaville commented:
> bors r-
Ok, I'll look at it tomorrow.

winksaville committed:
Fix failing doppelganger_protection test

Made `tests/vars.env` identical to `local_testnet/vars.env` except for
adding `VC_ARGS= --enable-doppelganger-protection ".

This was needed because of the renaming of NODE_COUNT to BN_COUNT.

winksaville commented:
> The doppelganger protection tests are failing due to the BN_COUNT rename. We can merge once that's fixed, but I'm going to drop this PR from the current batch (and the v2.0.0-rc.0)

Fixed!

michaelsproul reviewed (approved):
Great. Will merge this after v2.0.0-rc.0 merges.

michaelsproul commented:
bors r+
